src/circuit.rs

```diff
@@ -50,13 +270,401 @@ impl plonk::Circuit<pallas::Base> for Circuit {
+            layouter.assign_region(
+                || "v_old - v_new = magnitude * sign",
+                |mut region| {
+                    copy(&mut region, || "v_old", config.advices[0], 0, &v_old)?;
+                    copy(&mut region, || "v_new", config.advices[1], 0, &v_new)?;
+                    let (magnitude, sign) = v_net;
+                    copy(
+                        &mut region,
+                        || "v_net magnitude",
+                        config.advices[2],
+                        0,
+                        &magnitude,
+                    )?;
+                    copy(&mut region, || "v_net sign", config.advices[3], 0, &sign)?;
+
+                    config.q_v_net.enable(&mut region, 0)
+                },
+            )?;
+
+            // commitment = [v_net] ValueCommitV
+            let (commitment, _) = {
+                let value_commit_v = ValueCommitV::get();
+                let value_commit_v = FixedPointShort::from_inner(ecc_chip.clone(), value_commit_v);
+                value_commit_v.mul(layouter.namespace(|| "[v_net] ValueCommitV"), v_net)?
+            };
+
+            // blind = [rcv] ValueCommitR
+            let (blind, _) = {
+                let rcv = self.rcv.as_ref().map(|rcv| **rcv);
+                let value_commit_r = OrchardFixedBasesFull::ValueCommitR;
+                let value_commit_r = FixedPoint::from_inner(ecc_chip.clone(), value_commit_r);
+
+                // [rcv] ValueCommitR
+                value_commit_r.mul(layouter.namespace(|| "[rcv] ValueCommitR"), rcv)?
+            };
+
+            // [v_net] ValueCommitV + [rcv] ValueCommitR
+            commitment.add(layouter.namespace(|| "cv_net"), &blind)?
+        };
+
+        // Nullifier integrity
+        // TODO: constrain to equal public input nf_old
+        let nf_old = {
+            // nk_rho_old = poseidon_hash(nk, rho_old)
+            let nk_rho_old = {
+                let message = [nk, rho_old];
+
+                let poseidon_message = layouter.assign_region(
+                    || "load message",
+                    |mut region| {
+                        let mut message_word = |i: usize| {
+                            let value = message[i].value();
+                            let var = region.assign_advice(
+                                || format!("load message_{}", i),
+                                config.poseidon_config.state[i],
+                                0,
+                                || value.ok_or(plonk::Error::SynthesisError),
+                            )?;
+                            region.constrain_equal(var, message[i].cell())?;
+                            Ok(Word::<_, _, poseidon::OrchardNullifier, 3, 2> {
+                                inner: StateWord::new(var, value),
+                            })
+                        };
+
+                        Ok([message_word(0)?, message_word(1)?])
+                    },
+                )?;
```

Comment:
This seems a bit clunky, and should probably be handled by a function in the Poseidon code.